### ISSUE TITLE
Cryptic error when making a mistake in params definition

### ISSUE THREAD

**Issue by katafrakt** [hanami/hanami#1252]
Hi. I encountered a very cryptic error message when making an error in `params`, like this (`filed` instead of `filled`):

```ruby
  params do
    required(:name).filed(:str?)
  end
```

When booting the app and hitting a controller with an error, the message is:

```
Puma caught this error: wrong constant name #<Dry

        _scopes << _scopes.last.const_get(scope)
                               ^^^^^^^^^^ (NameError)
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:30:in `const_get'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:30:in `block in scopes'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:29:in `each'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:29:in `inject'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:29:in `scopes'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:21:in `class_names'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:14:in `corrections'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/errors.rb:99:in `initialize'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/loader.rb:70:in `new'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/loader.rb:70:in `rescue in constant'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/loader.rb:64:in `constant'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/loader/autoloading.rb:17:in `require!'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/loader.rb:47:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/component.rb:64:in `instance'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/auto_registrar.rb:34:in `block (2 levels) in call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-core-1.0.0/lib/dry/core/container/item/callable.rb:16:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-core-1.0.0/lib/dry/core/container/resolver.rb:36:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-core-1.0.0/lib/dry/core/container/mixin.rb:132:in `resolve'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-system-1.0.0.rc1/lib/dry/system/container.rb:493:in `resolve'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/hanami-2.0.0.rc1/lib/hanami/slice.rb:599:in `resolve'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/hanami-2.0.0.rc1/lib/hanami/slice/routing/resolver.rb:74:in `block in resolve_slice_action'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/hanami-router-2.0.0.rc1/lib/hanami/router.rb:107:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-monitor-1.0.0/lib/dry/monitor/rack/middleware.rb:35:in `block in call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-monitor-1.0.0/lib/dry/monitor/clock.rb:10:in `measure'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-monitor-1.0.0/lib/dry/monitor/rack/middleware.rb:35:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/hanami-router-2.0.0.rc1/lib/hanami/middleware/app.rb:40:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/hanami-2.0.0.rc1/lib/hanami/slice.rb:758:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/configuration.rb:268:in `call'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:89:in `block in handle_request'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:343:in `with_force_shutdown'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:88:in `handle_request'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:430:in `process_client'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:233:in `block in run'
/home/katafrakt/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:150:in `block in spawn_thread'
```

The repo with reproduction is here: https://github.com/katafrakt/repro_hanami_params_error (just start the server and go to `localhost:2300`)

**Comment by timriley** [comment]
Thanks for sharing this, @katafrakt!

FWIW, I tested this another way, and it's revealed the information I would have expected/wanted to see:

```ruby
testing_things[development]> app["actions.home"]
NameError: wrong constant name #<Dry

        _scopes << _scopes.last.const_get(scope)
                               ^^^^^^^^^^
from /Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/3.1.0/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb:30:in `const_get'
Caused by NoMethodError: undefined method `filed' for #<Dry::Schema::Macros::Required:0x0000000107070eb0 @name=:name, @compiler=#<Dry::Schema::Compiler:0x00000001070939b0 @predicates=#<Dry::Schema::PredicateRegistry:0x0000000107093aa0 @predicates=Dry::Logic::Predicates>>, @trace=#<Dry::Schema::Trace compiler=#<Dry::Schema::Compiler:0x000000010706bd48 @predicates=#<Dry::Schema::PredicateRegistry:0x000000010706bdc0 @predicates=Dry::Logic::Predicates>> captures=[]>, @schema_dsl=#<Dry::Schema::DSL:0x00000001070981b8 @compiler=#<Dry::Schema::Compiler:0x00000001070939b0 @predicates=#<Dry::Schema::PredicateRegistry:0x0000000107093aa0 @predicates=Dry::Logic::Predicates>>, @processor_type=Dry::Schema::Params, @macros=[#<Dry::Schema::Macros::Required:0x0000000107070eb0 ...>], @types={:name=>#<Dry::Types[Any]>}, @parents=[#<Dry::Schema::Params keys=["_csrf_token"] rules={:_csrf_token=>"key?(:_csrf_token) THEN key[_csrf_token](filled? AND str?)"}>], @config=#<Dry::Validation::Config messages={:top_namespace=>"dry_validation", :load_paths=>#<Set: {#<Pathname:/Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-schema-1.12.0/config/errors.yml>, #<Pathname:/Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-validation-1.10.0/config/errors.yml>}>, :backend=>:yaml, :namespace=>nil, :default_locale=>nil} macros=#<Dry::Validation::Macros::Container:0x00000001070990e0 @_container={}> types=Dry::Types validate_keys=false predicates=#<Dry::Schema::PredicateRegistry:0x0000000106c84158 @predicates=Dry::Logic::Predicates>>, @steps=#<Dry::Schema::ProcessorSteps:0x0000000107093050 @steps={}, @before_steps={}, @after_steps={}>, @path=#<Dry::Schema::Path keys=[]>, @filter_schema_dsl=#<Dry::Schema::DSL:0x0000000107078930 @compiler=#<Dry::Schema::Compiler:0x0000000107078318 @predicates=#<Dry::Schema::PredicateRegistry:0x0000000107078368 @predicates=Dry::Logic::Predicates>>, @processor_type=Dry::Schema::Params, @macros=[], @types={}, @parents=[], @config=#<Dry::Validation::Config messages={:top_namespace=>"dry_validation", :load_paths=>#<Set: {#<Pathname:/Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-schema-1.12.0/config/errors.yml>, #<Pathname:/Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-validation-1.10.0/config/errors.yml>}>, :backend=>:yaml, :namespace=>nil, :default_locale=>nil} macros=#<Dry::Validation::Macros::Container:0x00000001070990e0 @_container={}> types=Dry::Types validate_keys=false predicates=#<Dry::Schema::PredicateRegistry:0x0000000106c84158 @predicates=Dry::Logic::Predicates>>, @steps=#<Dry::Schema::ProcessorSteps:0x0000000107073e30 @steps={}, @before_steps={}, @after_steps={}>, @path=#<Dry::Schema::Path keys=[]>>>, @chain=true, @predicate_inferrer=#<Dry::Schema::PredicateInferrer:0x000000010706b910 @compiler=#<Dry::Schema::PredicateInferrer::Compiler:0x000000010706b578 @registry=#<Dry::Schema::PredicateRegistry:0x0000000107093aa0 @predicates=Dry::Logic::Predicates>>>, @primitive_inferrer=#<Dry::Schema::PrimitiveInferrer:0x000000010706b230 @compiler=#<Dry::Schema::PrimitiveInferrer::Compiler:0x000000010706ac40>>, @filter_schema_dsl=#<Dry::Schema::DSL:0x0000000107078930 @compiler=#<Dry::Schema::Compiler:0x0000000107078318 @predicates=#<Dry::Schema::PredicateRegistry:0x0000000107078368 @predicates=Dry::Logic::Predicates>>, @processor_type=Dry::Schema::Params, @macros=[], @types={}, @parents=[], @config=#<Dry::Validation::Config messages={:top_namespace=>"dry_validation", :load_paths=>#<Set: {#<Pathname:/Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-schema-1.12.0/config/errors.yml>, #<Pathname:/Users/tim/.asdf/installs/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/dry-validation-1.10.0/config/errors.yml>}>, :backend=>:yaml, :namespace=>nil, :default_locale=>nil} macros=#<Dry::Validation::Macros::Container:0x00000001070990e0 @_container={}> types=Dry::Types validate_keys=false predicates=#<Dry::Schema::PredicateRegistry:0x0000000106c84158 @predicates=Dry::Logic::Predicates>>, @steps=#<Dry::Schema::ProcessorSteps:0x0000000107073e30 @steps={}, @before_steps={}, @after_steps={}>, @path=#<Dry::Schema::Path keys=[]>>>

    required(:name).filed(:str?)
                   ^^^^^^
Did you mean?  filled
               filter
from /Users/tim/Source/github.com/katafrakt/repro_hanami_params_error/app/actions/home.rb:3:in `block in <class:Home>'
```

I think the error is getting swallowed up by dry-system's `DidYouMean` integration when resolving components, which is intended to provide helpful error messages, but in this case is doing the complete opposite!

I'll take a look into what's going on there, but if there's no obvious fix, my inclination at this point would be to disable the DidYouMean integration until we can ensure it always helps, never hinders. This example will be invaluable for testing it, thank you so much for sharing!

**Comment by cllns** [comment]
@timriley I'm looking into this. I added [the DidYouMean integration in dry-system](https://github.com/dry-rb/dry-system/pull/217) so I have the best chance of wrangling this. If it's too much, yes we should disable it but I hope we can figure it out!

**Comment by timriley** [comment]
Thank you @cllns! I'll leave with with you for a few days then. 

IIRC I've also been confused by obtuse results from the DidYouMean integration once or twice, but unfortunately I can't recall the details. Luckily we have this great example to work from now.

**Comment by cllns** [comment]
I've figured out the problem and a fix.

dry-system is currently catching __all__ `NameError`s that happen, and `NoMethodError` (for `.filed`) _is_ a `NameError`. Instead we only want to do our special DidYouMean integration for the one specific constant we're trying to load, none that happen within the loaded constant itself.

Just need to figure out how to write some good tests for it because this loading stuff and messing with errors is tricky and high impact 😬

**Comment by timriley** [comment]
Ah yes, this sounds like the crux of it!

**Comment by solnic** [comment]
I can confirm that this now works as expected using dry-system from main. So, it'll be fixed in 1.0.0.
